AdvancedDLSupport/ImplementationGenerators/Wrappers/DelegateWrapper.cs

```diff
@@ -0,0 +1,229 @@
+
+            il.Emit(OpCodes.Call, marshalPointerToDel);
+            il.Emit(OpCodes.Ret);
+
+            il.MarkLabel(retNullLabel);
+
+            il.Emit(OpCodes.Ldnull);
+        }
+
+        /// <summary>
+        /// Gets the delegate lifetime that the parameter with the given attributes should be marshalled with.
+        /// </summary>
+        /// <param name="customAttributes">The custom attributes applied to the parameter.</param>
+        /// <returns>The delegate lifetime.</returns>
+        [Pure]
+        private DelegateLifetime GetParameterDelegateLifetime([NotNull, ItemNotNull] IEnumerable<CustomAttributeData> customAttributes)
+        {
+            var lifetimeAttribute = customAttributes.FirstOrDefault
+            (
+                a =>
+                    a.AttributeType == typeof(DelegateLifetimeAttribute)
+            );
+
+            if (lifetimeAttribute is null)
+            {
+                // Default to keeping delegates alive
+                return DelegateLifetime.Persistent;
+            }
+
+            return lifetimeAttribute.ToInstance<DelegateLifetimeAttribute>().Lifetime;
+        }
+
+        /// <inheritdoc/>
+        public override IntrospectiveMethodInfo GeneratePassthroughDefinition
+        (
+            PipelineWorkUnit<IntrospectiveMethodInfo> workUnit
+        )
+        {
+            var definition = workUnit.Definition;
+
+            var newReturnType = GetParameterPassthroughType(definition.ReturnType, GetParameterDelegateLifetime(definition.ReturnParameterCustomAttributes));
+            var newParameterTypes = definition.ParameterTypes.Select
+            (
+                (t, i) => GetParameterPassthroughType(t, GetParameterDelegateLifetime(definition.ParameterCustomAttributes[i]))
+            ).ToArray();
+
+            var passthroughMethod = TargetType.DefineMethod
+            (
+                $"{workUnit.GetUniqueBaseMemberName()}_wrapped",
+                MethodAttributes.Private | MethodAttributes.Virtual | MethodAttributes.HideBySig,
+                CallingConventions.Standard,
+                newReturnType,
+                newParameterTypes
+            );
+
+            passthroughMethod.ApplyCustomAttributesFrom(definition, newReturnType, newParameterTypes);
+
+            return new IntrospectiveMethodInfo
+            (
+                passthroughMethod,
+                newReturnType,
+                newParameterTypes,
+                definition.MetadataType,
+                definition
+            );
+        }
+
+        /// <summary>
+        /// Gets the type that the parameter type should be passed through as.
+        /// </summary>
+        /// <param name="originalType">The original type.</param>
+        /// <returns>The passed-through type.</returns>
+        [NotNull]
+        private Type GetParameterPassthroughType([NotNull] Type originalType, DelegateLifetime lifetime)
```

Comment:
Should we not always pass through an IntPtr, regardless of lifetime settings? Marshalling delegates through `GetFunctionPointerForDelegate` is required no matter what kind of lifetime the user chooses because of relocation issues.

Reply:
Indeed. Do we want that for UserManaged as well?

Reply:
Yeah, I think so.